design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+  keep in mind that having at least a function defined would invalidate the
+  assumption that rendered Managed Resources should be valid according to the
+  schema. Moreover, taking them into consideration during the validation phase
+  would mean adding another external dependency, which could both reduce
+  reliability and add latency.
+- `environment`: contains a list of `EnvironmentConfigs` that can be referenced
+  by specific types of patches and a list of`EnvironmentPatches` . Also an
+  `alpha` feature, similar implications to what already said for `functions`,
+  would require additional information and can invalidate the assumption that
+  we render valid Composed Resources if defined.
+- `writeConnectionSecretsToNamespace` : the name of the namespace the `Secret`
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. The CRDs of the Managed Resources defined as `base` and traversed by
+   `patches` might not yet have been created when the Composition is being
+   validated.
+2. The CRD of the Composite Resources specified by `CompositeTypeRef` and
+   traversed by `patches` might not yet have been created when the Composition
+   is being validated.
+3. We do not have all the required CRDs locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+In order to address the possible absence of Managed Resources' CRDs, Composite
+Resources' CRDs or any other required external resource at validation time we
+should allow explicitly setting one of 2 modes, `strict` and `loose`(default) .
+These should be configurable through an annotation, e.g.
+`crossplane.io/composition-validation-mode`, on a Composition or maybe even
+globally via a flag, and would respectively imply that any missing resource
+should result in a direct rejection or a possibly incomplete validation.
```

Comment:
Is there a reason to prefer an annotation to a 'real' field on the Composition (e.g. `spec.validation.policy` or similar)?

Reply:
I'd go with an annotation for the first iterations as it allows us not have to change the APIs and let us change our mind more easily, we could think of moving it to the spec at a later stage once the UX settles to make it more "official" and easily discoverable.

Reply:
That sounds good to me.

I do recall though that upstream Kubernetes moved away from the practice you described for some reason. Might be worth checking why.